pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,777 @@
+    >>> tape.get_parameters()
+    [0.56]
+    >>> tape.get_parameters(free_only=False)
+    [0.56, 0.543, 0.133]
+
+    Trainable parameters are taken into account when calculating the Jacobian,
+    avoiding unnecessary calculations:
+
+    >>> tape.jacobian(dev)
+    [[-0.45478169]]
+    """
+
+    _cast = staticmethod(np.array)
+
+    def __init__(self, name=None):
+        super().__init__()
+        self.name = name
+
+        self._prep = []
+        """list[~.Operation]: Tape state preparations."""
+
+        self._ops = []
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._obs = []
+        """list[tuple[~.MeasurementProcess, ~.Observable]]: measurement processes and
+        coresponding observables recorded by the tape."""
+
+        self._par_info = {}
+        """dict[int, dict[str, Operation or int]]: Parameter information. Keys are
+        parameter indices (in the order they appear on the tape), and values are a
+        dictionary containing the corresponding operation and operation parameter index."""
+
+        self._trainable_params = set()
+        self._graph = None
+        self._output_dim = 0
+
+        self.wires = qml.wires.Wires([])
+        self.num_wires = 0
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: wires={self.wires}, params={self.num_params}>"
+
+    def __enter__(self):
+        if not QueuingContext.recording():
+            # if the tape is the first active queuing context
+            # monkeypatch the operations to support the new queuing context
+            with contextlib.ExitStack() as stack:
+                for mock in mock_operations():
+                    stack.enter_context(mock)
+                self._stack = stack.pop_all()
+
+        QueuingContext.append(self)
+        return super().__enter__()
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        super().__exit__(exception_type, exception_value, traceback)
+
+        if not QueuingContext.recording():
+            # remove the monkeypatching
+            self._stack.__exit__(exception_type, exception_value, traceback)
+
+        self._process_queue()
+
+    @property
+    def interface(self):
+        """str, None: automatic differentiation interface used by the quantum tape (if any)"""
+        return None
+
+    # ========================================================
+    # construction methods
+    # ========================================================
+
+    def _process_queue(self):
+        """Process the annotated queue, creating a list of quantum
+        operations and measurement processes.
+
+        This method sets the following attributes:
+
+        * ``_ops``
+        * ``_obs``
+        * ``_par_info``
+        * ``_output_dim``
+        * ``_trainable_params``
+        * ``is_sampled``
```

Comment:
A quick comment: In the capabilities dictionary (where names are supposed to be *very* explicit because they can be queried by ecosystem app developers in future) I recently changed to using the term "supports_finite_shots". That was quite an aha moment, because 
a) what we really mean by sampled is that a device is run with finite shots
b) this nomenclature answers immediately the question of *what* is sampled
b) I expect after the upcoming shots ADR we decide to move all the "analytic or not" logic into the shots argument anyways

So maybe use ``finite_shots`` here?

Reply:
Oh, this refers to the presence of `qml.sample()` measurements; it is independent of whether the device is run in 'analytic' mode or not. The reason we have to check for this is that the presence of sampling as output changes the output type --- we now have to return object numpy arrays.

I also don't think I can change it (at the moment), if I recall there are some assumptions elsewhere in the codebase that assume this attribute name!